### PR TITLE
Authenticate As for Cloudcare

### PR DESCRIPTION
**Pull request by benrudolph**
@czue @NoahCarnahan 
cc: @millerdev 

What I was thinking for allowing superusers to authenticate as for cloudcare. Not sure if this is the best approach, so I put up a WIP to get feedback

### CODE REVIEW

File: corehq/apps/cloudcare/middleware.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+from django.contrib import messages
+from django.contrib.auth.models import User
+from django.contrib.auth import login
+
+
+class CloudcareMiddleware(object):
+
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        auth_as = request.GET.get('auth_as', None)
+        view_name = '.'.join((view_func.__module__, view_func.__name__))
```
Comment (dimagimon) [inline corehq/apps/cloudcare/middleware.py]: F841 local variable 'view_name' is assigned to but never used

File: corehq/apps/cloudcare/tests/test_middleware.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+from django.test import TestCase
```
Comment (dimagimon) [inline corehq/apps/cloudcare/tests/test_middleware.py]: F401 'TestCase' imported but unused

File: corehq/apps/cloudcare/tests/test_middleware.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+from django.core.urlresolvers import reverse
+
+from corehq.apps.accounting.tests.base_tests import BaseAccountingTest
+from corehq.apps.cloudcare.middleware import CloudcareMiddleware
+from corehq.apps.users.models import WebUser, CommCareUser
+
+
+class CloudcareMiddleware(BaseAccountingTest):
```
Comment (dimagimon) [inline corehq/apps/cloudcare/tests/test_middleware.py]: F811 redefinition of unused 'CloudcareMiddleware' from line 5

File: corehq/apps/cloudcare/tests/test_middleware.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        Ensures that using the URL param ?auth_as=<username> successfully
+        sets the user in the cloudcare middleware
+        """
+        username = '{username}@{domain}.commcarehq.org'.format(
+            username='muggle',
+            domain=self.domain,
+        )
+        password = 'I solemnly swear that I am up to no good'
+
+        muggle = CommCareUser.create(self.domain, username, password)
```
Comment (dimagimon) [inline corehq/apps/cloudcare/tests/test_middleware.py]: F841 local variable 'muggle' is assigned to but never used

File: corehq/apps/cloudcare/middleware.py (hunk 1 above)
Comment (benrudolph) [inline corehq/apps/cloudcare/middleware.py]: This isn't necessary in the current state, but I was going to use it to check if the view belongs to the cloudcare app.

**Comment by benrudolph** [comment]
(will clean up dimagimon if you think this is a good path to go down)

**Comment by NoahCarnahan** [comment]
This seems like a fine approach to me so far. One thing you should be conscious of is that the order of the middleware will matter. We have other middlewares that touch the `request.user` (like `corehq.apps.users.middleware.UsersMiddleware`), so you should make sure this middleware interacts with those as expected.

File: corehq/apps/cloudcare/tests/test_middleware.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    def test_auth_as(self):
+        """
+        Ensures that using the URL param ?auth_as=<username> successfully
+        sets the user in the cloudcare middleware
+        """
+        username = '{username}@{domain}.commcarehq.org'.format(
+            username='muggle',
+            domain=self.domain,
+        )
+        password = 'I solemnly swear that I am up to no good'
```
Comment (snopoke) [inline corehq/apps/cloudcare/tests/test_middleware.py]: I don't think muggles can do magic :trollface:

File: corehq/apps/cloudcare/middleware.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        if request.user and request.user.is_superuser and auth_as:
+            try:
+                request.user = User.objects.get(username=auth_as)
```
Comment (czue) [inline corehq/apps/cloudcare/middleware.py]: tiny but it'd be nice to just have to pass in the short part of the id and have HQ generate the domain-specific username for you (not sure how hard it is to get access to the domain or if that's even possible in all views)

File: corehq/apps/cloudcare/middleware.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+            except User.DoesNotExist:
+                messages.warning(request, u'Unable to find user: {} as'.format(auth_as))
+            else:
+                # http://stackoverflow.com/a/2787747/835696
+                # This allows us to bypass the authenticate call
+                request.user.backend = 'django.contrib.auth.backends.ModelBackend'
+                login(request, request.user)
```
Comment (czue) [inline corehq/apps/cloudcare/middleware.py]: is the idea that you do this once and then you "become" that user for the remainder of the session? or does this not persist across requests?

**Comment by czue** [comment]
I'm not sure off the top of my head whether middleware makes more sense than something else. another interesting approach would be to have a single view you can hit that "logs you in" as that user, and then you can go visit whatever URL you want. 

Thinking about all the different places this might be affected (ajax calls, touchforms api callbacks, etc.) it does seem like just getting a session as a different user is going to be easier than pumping the url param through every request

**Comment by benrudolph** [comment]
I like the idea of having a single view. seems less risky to me. the reason why it's nice to have it in the middleware is for the exact reasons that @NoahCarnahan was pointing out. we do some stuff in the middlewares that we'd have to duplicate in the view

File: corehq/apps/cloudcare/middleware.py (hunk 6 above)
Comment (benrudolph) [inline corehq/apps/cloudcare/middleware.py]: :+1:

**Comment by NoahCarnahan** [comment]
I actually think you might not have to duplicate the middleware stuff Ben. Let's say you have a from, only available to super users, for using hq as someone else. It has a field for username, but no password field. The view that processes this form would be just like the normal login form, but doesn't check the password. This would change the user of the session and all that, and wouldn't require changing any other views I don't think. (This might be what @czue was suggesting, or a little different)

**Comment by czue** [comment]
that's basically what i was suggesting but with a more thought through UI

**Comment by benrudolph** [comment]
That makes more sense to me. thanks for feedback. closing until i rework it.

**Pull request closed without merging**